### PR TITLE
Normal traction BC

### PR DESCRIPTION
**Pull request by sytuannguyen**
This PR adds a possibility for applying normal traction boundary condition. This is needed for applying pressure on a curved boundary such as an elastic/poroelastic borehole with inner pressure. This example is also added to the integratedTest.

The input of this BC is designed as follow:
image

with usual time function

image

Below is the results of the borehole example

image

The validation against analytical results is shown below:

image

Finer mesh than that in the integrated test is considered for the validation

image

### CODE REVIEW

**Review by TotoGaz — COMMENTED** [review]
I am really surprised to see such a generic code in a specialized file (for `SolidMechanicsLagrangianFEM`).
I also think that this code needs refactoring above the range of your PR.

File: src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp [hunk 1]
```diff
@@ -712,7 +712,8 @@ void SolidMechanicsLagrangianFEM::crsApplyTractionBC( real64 const time,
   FaceManager const & faceManager = *domain.getMeshBody( 0 )->getMeshLevel( 0 )->getFaceManager();
   NodeManager const & nodeManager = *domain.getMeshBody( 0 )->getMeshLevel( 0 )->getNodeManager();
 
-  arrayView1d< real64 const > const faceArea  = faceManager.getReference< real64_array >( "faceArea" );
+  arrayView1d< real64 const > const faceArea  = faceManager.faceArea();
+  arrayView2d< real64 const > const faceNormal  = faceManager.faceNormal();
```
[the remark this answers is not in the document]
Comment (sytuannguyen) [reply src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp:716]: This variable name was used elwhere, then I decide to reuse it.

File: src/coreComponents/physicsSolvers/solidMechanics/integratedTests/borehole.xml [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+      fieldName="TotalDisplacement"
+      component="2"
+      scale="0.0"
+      setNames="{ zneg, zpos }"/>
+
+    <FieldSpecification
+      name="borehole stress"
+      objectPath="faceManager"
+      fieldName="Traction"
+      component="0"
```
[the remark this answers is not in the document]
Comment (sytuannguyen) [reply src/coreComponents/physicsSolvers/solidMechanics/integratedTests/borehole.xml]: Thank you to point this out, it was changed for testing other boundary conditions. I changed it back to -1.

File: src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp [hunk 3]
```diff
@@ -752,10 +753,24 @@ void SolidMechanicsLagrangianFEM::crsApplyTractionBC( real64 const time,
         localIndex const numNodes = faceToNodeMap.sizeOfArray( kf );
         for( localIndex a=0; a<numNodes; ++a )
         {
-          localIndex const dof = blockLocalDofNumber[ faceToNodeMap( kf, a ) ] + component - dofRankOffset;
-          if( dof < 0 || dof >= localRhs.size() )
-            continue;
-          RAJA::atomicAdd< parallelDeviceAtomic >( &localRhs[ dof ], value * faceArea[kf] / numNodes );
+          if( component != -1 )
```
[the remark this answers is not in the document]
Comment (sytuannguyen) [reply src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp:756]: I prefer this simple option for this task instead of complexify both the code and the input file. How do you think about this @rrsettgast ?

**Comment by sytuannguyen** [comment]
> I am really surprised to see such a generic code in a specialized file (for `SolidMechanicsLagrangianFEM`).
> I also think that this code needs refactoring above the range of your PR.

I  agree that the solid solver merit some clearning/refractoring and both traction and chombo pressure boundary conditions should be moved to field specification manager/base. Let's do it in other PRs to keep this one forcus on the main objective.

File: src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (TotoGaz) [reply src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp:756]: Hi @sytuannguyen :)

Simple is a relative concept: while the code is surely simple to write, it's less simple to read, maintain or understand.
When I read `if( component != -1 )` I am unable to understand what this means without further investigations: I do not consider this code simple for this reason.

Furthermore, this value is directly exposed to the end-user up to the heart of BC settings, so if I want to modify something in the code someday I will have a lot to modify (to the users input files in fact). Worse, I'll need to be aware that there is a constant value here, making refactoring even more dangerous.

Also, the code of this function (and it's not because of you 😉  ) is duplicated (even quadrupled in some sense), so we need to be careful about this in the future.

Maybe/probably it's not to you to fix all of this, but it's worth mentioning.

[the remark this answers is not in the document]
Comment (sytuannguyen) [reply src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp:756]: I agree with your points. For instant I will add some comments to clarify this flag. Let's make this BC work first for the actual requirement in the wellbore applications. Then we will optimize it together with the whole solid solver in future PRs.

File: src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp [hunk 4]
```diff
@@ -752,10 +753,29 @@ void SolidMechanicsLagrangianFEM::crsApplyTractionBC( real64 const time,
...
         for( localIndex a=0; a<numNodes; ++a )
         {
-          localIndex const dof = blockLocalDofNumber[ faceToNodeMap( kf, a ) ] + component - dofRankOffset;
-          if( dof < 0 || dof >= localRhs.size() )
-            continue;
-          RAJA::atomicAdd< parallelDeviceAtomic >( &localRhs[ dof ], value * faceArea[kf] / numNodes );
+
+          // The flag component = 0, 1 and 2 are for the traction in the three orthogonal directions
+          // in the global cartesian coordinate system  
+          // The flag component = -1 is for the normal traction
```
Comment (TotoGaz) [inline src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp:759]: This is probably not the correct location where you should put your comment.
The value comes from `FieldSpecificationBase::getComponent` So this is where your comment should go.
E.g.: why didn't you copy/paste your comment below (line 796) too?

You should improve the user message [here](https://github.com/GEOSX/GEOSX/blob/b204d2c528d59533929bca7f7a06b4946b8b2da2/src/coreComponents/managers/FieldSpecification/FieldSpecificationBase.cpp#L48) too.

**Review by TotoGaz — COMMENTED** [review]
We can let it go after the comment is updated, but I'm afraid we'll have maintenance issues with this BC stuff someday.

File: src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp (hunk 4 above)
Comment (sytuannguyen) [reply src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp]: Done!

File: src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (rrsettgast) [reply src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp:761]: @CusiniM How are you applying pressures to the embedded fractures?

[the remark this answers is not in the document]
Comment (CusiniM) [reply src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp:761]: Hmm. I have registered a field on the `embeddedSurfaceSubRegion` called `Traction` and that field is updated by a constitutive model (the contact relationship) based on the primary variables. In my case it depends on the pressure and on the fracture aperture. Being a registered field I can also set it as a b.c. in the fieldSpecifications. This is not in develop yet but it is all in https://github.com/GEOSX/GEOSX/pull/1229/files

**Comment by CusiniM** [comment]
@sytuannguyen, you may want to add the validation case to the benchmarks if the mesh is too fine for it to be an integrated test.

**Comment by sytuannguyen** [comment]
> @sytuannguyen, you may want to add the validation case to the benchmarks if the mesh is too fine for it to be an integrated test.

@CusiniM Thank you for your suggestion. The test using fine mesh is now added to solid mechanic benchmarks.

**Comment by sytuannguyen** [comment]
@rrsettgast  plz jump in if you are not OK with the flag -1 for normal traction boundary condition.

**Comment by sytuannguyen** [comment]
@TotoGaz @rrsettgast @CusiniM I will merge this PR to #1338 because this is needed for testing the wellbore problem. So I will close this.
